### ISSUE TITLE
perf: explore a two-stage segmentation + per-word POS tagging architecture

### ISSUE THREAD

**Issue by mosuka** [mosuka/litsea#147]
## Background

The #143 campaign (PR #144) ported segment_with_pos() to the packed two-pass scorer: 558K -> ~1.9M chars/s extrapolated on the external benchmark (13th -> ~8th place). The f32 follow-up (#145) measured zero benefit and was closed: the workload is latency-bound, and element-width reduction does not reduce the number of random accesses.

The remaining cost is architectural: the joint model scores **17 classes at every character position**. vaporetto and KyTea avoid this with a two-stage design — a binary boundary classifier (litsea already runs this class of model at ~11M chars/s) followed by a per-word tag classifier evaluated once per word (~1.7 chars/word in Japanese), typically with per-word candidate restriction. This is the only remaining route toward vaporetto-class throughput for the POS path.

## Goal

Design-first exploration, then implementation if the design holds:

1. **Design doc**: two-stage pipeline shape for litsea — boundary model (reuse the AdaBoost path or a binary perceptron), word-level feature template for the tagger, candidate restriction strategy, model file format, and how training corpora map onto the two stages.
2. **Quality evaluation**: joint vs two-stage on held-out data. The joint model's boundary decisions are informed by POS signal; splitting may change segmentation quality, not just tagging quality. Both must be measured before committing.
3. **Implementation** behind the existing public API (`segment_with_pos()` signature unchanged), with new model files.

## Non-goals / constraints

- This is NOT an output-equivalent optimization: predictions will differ from the joint model. Existing joint-model files must keep working (the two-stage path should be additive, selected by model type).
- Requires retraining and new bundled models for ja/zh/ko.

## Acceptance criteria

- Design doc reviewed before implementation starts (scale assumptions and priority metric confirmed with the maintainer, per past lessons).
- Segmentation and POS quality parity gates defined from measured baseline variance, not aspirational numbers.
- Benchmark target: multi-x improvement over the current ~1.9M chars/s equivalent; the exact target is set in the design doc after prototyping the per-word tagger cost.

## Tasks

- [x] Investigation: corpus/feature design for per-word tagging; survey vaporetto/KyTea tag-prediction details (verify against their actual implementations, not summaries)
- [x] Design doc + maintainer review
- [x] Prototype + quality evaluation (joint vs two-stage)
- [x] Go/no-go decision — **GO** (2026-08-14): quality beats joint across all measured configs; speed gate updated to >=2.5x required (demonstrated), 3x stretch
- [x] Implementation, training pipeline, bundled models, docs — child issues:
  - [x] #166 model file format and loader (litsea-two-stage v1) — merged in #170
  - [x] #167 runtime: per-word tagger behind segment_with_pos() — merged in #171
  - [x] #168 training pipeline and CLI — merged in #172
  - [x] #169 bundled models, cross-language validation, docs and benches — merged in #173

Spin-off (independent of this issue): #165 collapsed binary-AP bundled segmentation models.

## References

- #143 / PR #144 (packed two-pass port), #145 (f32 negative result)
- Investigation data: 17 classes, 96,487 features, avg 3.27 non-zero classes/feature (japanese_pos.model)

**Comment by mosuka** [comment]
## Investigation report: two-stage segmentation + per-word POS tagging

Investigation phase is complete. Per the acceptance criteria, vaporetto and KyTea were surveyed against their **actual source code** (not summaries), alongside a full audit of litsea's current state. Detailed reports with file:line citations are kept locally; this comment is the consolidated summary.

Surveyed targets:

- vaporetto v0.6.5 (daac-tools/vaporetto, commit `15baaa0`), Rust
- KyTea (neubig/kytea, commit `73a94c4`), C++
- litsea main (v0.10.0)

### How the reference implementations do tag prediction

| Aspect | vaporetto | KyTea |
|---|---|---|
| Stage-2 unit | per-word mini models keyed by surface (`HashMap<String, TagPredictor>`) | per-word ("local") models by default; optional single "global" classifier (`-global`, recommended for POS in usage text) |
| Candidate restriction | hard-restricted to tags observed for that surface in training data + dictionary | local: same hard restriction; global: dictionary membership as soft features (`D{d}T{t}` / `UNK`) |
| Single-candidate words | no classifier trained; tag assigned with zero scoring | same (local path): NULL model → first candidate immediately |
| Unknown words | **no tag at all** (`None`; no fallback model) | global model classifies them normally; local-only path falls back to a subword generative model (meant for pronunciation) or default tag `"UNK"` |
| Stage-2 features | char/type n-grams that fully contain the word and extend it by 1..n chars; surface itself is not a feature (per-word model split + bias encodes it) | char/type n-grams of the 3 chars left/right of the word (word-internal chars excluded) + surface self feature + dict features (latter two global-only) |
| Classifier | liblinear, int16-quantized weights | liblinear (default L2R_L2LOSS_SVC_DUAL, one-vs-rest), int16-quantized |
| Runtime cost per word | 1 surface hash lookup + O(window) PMA-state probes + n_class vector add; context features ride on the boundary PMA pass (matched state ids are cached, no re-scan) | 1 dict trie lookup + 2 Aho-Corasick matches over ≤6 chars + hits × n_class int adds; unambiguous words finish after the dict lookup |

Key takeaways:

1. **Candidate restriction is the main cost lever in both.** Unambiguous words (the majority in real text) never invoke a classifier.
2. **Neither implementation scores all 17 classes for every word** — effective class counts are 1 to a few per word.
3. Neither has a strong answer for unknown-word tagging (vaporetto: none; KyTea: global model or default tag). litsea's `segment_with_pos()` contract (every word gets a `Upos`) has no direct precedent, so a two-layer design is needed: hard candidate lists for known surfaces + a global fallback classifier for unknown words.

### litsea current-state findings

- The 5-6x throughput gap (~10.6M vs ~1.8-1.9M chars/s) is confirmed to be purely the multiclass dimension: the POS path performs roughly the same number of probes/loads per char as the AdaBoost path, but every add is an n_classes-wide row plus argmax and an n×17 score-matrix traffic.
- No per-word machinery exists on the inference path today, but reusable parts are in place: word spans fall out of the greedy decode loop; `PackedPosModel`'s sparse class rows / presence bitset / argmax can be repurposed as the stage-2 scorer; `process_tokens` is a natural base for generating per-word training examples. The existing `word/POS` corpus format maps cleanly onto both stages.
- Model files have no type tag/magic, but both existing loaders were verified to reject a header line that is neither format (`InvalidData`), so **a new headered model format can be added without breaking existing joint/AdaBoost models** — satisfying the "additive, selected by model type" constraint.
- Quality baseline (held-out UD GSD test, reproducible via `litsea evaluate`): segmentation-only (AdaBoost) ja 91.48 / zh 77.56 / ko 99.91 Word F1; joint POS ja **96.56**/92.51, zh 90.52/81.18, ko 80.51/71.03 (Word F1 / Tagged F1).

### Main design risk

The essential trade-off is **segmentation quality, not speed**: on Japanese, the joint model's segmentation (96.56) beats the standalone AdaBoost model (91.48) by +5.08pt on the same training corpus — the POS signal demonstrably helps segmentation. A naive two-stage design inherits stage-1's 91.48. Mitigation candidates (to be evaluated in the design/prototype phase): stronger stage-1 training, a binary averaged-perceptron stage-1 (scalar weights, same speed class), or vaporetto-style dictionary features (larger scope).

First-order speed estimate: stage-1 runs at ~10.6M chars/s; with candidate restriction + single-candidate skip, stage-2 has no structural reason to dominate. A realistic target is 3-5x over the current ~1.9M chars/s, to be fixed in the design doc after prototyping (per the acceptance criteria).

### Next step

Confirm design premises with the maintainer (priority metric, segmentation-quality gate level, unknown-word tagging requirement, target workload), then write the design doc for review.

**Comment by mosuka** [comment]
## Design doc: two-stage segmentation + per-word POS tagging (for review)

Design premises were confirmed with the maintainer (2026-08-14):

| Item | Decision |
|---|---|
| Priority metric | **Throughput** (chars/s multiple; quality guarded by floor gates) |
| Segmentation quality gate | **Intermediate level** — fixed after prototyping stage-1 strengthening, from measured baseline variance (ja ballpark 94-95+; plain AdaBoost's 91.48 is not acceptable; full joint parity 96.56 is not required) |
| Unknown-word tagging | **Fallback classifier** (a global multiclass classifier always assigns a real tag) |
| Target workload | **Long-text throughput** (measured with `cargo bench -- external_corpus`) |

### Architecture

```text
input text
  → [stage-1] binary boundary classifier (existing 42/38 templates, scalar weights,
              same greedy decode structure as segment(); word spans fall out incrementally)
  → [stage-2] per-word tag decision, invoked as each word span is finalized:
       (a) lexicon lookup: surface → candidate tag list
           - single candidate → assign immediately (zero scoring)  … majority of tokens
           - multiple candidates → global classifier, candidate-masked argmax
       (b) not in lexicon (unknown) → global classifier, full 17-tag argmax
  → Vec<(String, Upos)>
```

Key decision: **no per-word models** (vaporetto / KyTea-local style). Instead a two-layer design: hard candidate restriction via a lexicon + ONE global multiclass classifier (KyTea-global style hybrid). Rationale: the unknown-word requirement needs a global classifier anyway; candidate restriction + single-candidate skip (the cost lever in both reference implementations) only needs the lexicon; and litsea's existing `AveragedPerceptron` + `PackedPosModel` machinery (sparse class rows / presence bitset / argmax) is directly reusable.

- **Stage-1**: existing template set, scalar weights (same speed class as `segment()`, ~10.6M chars/s ja). The +5pt segmentation-quality recovery is an explicit Phase 0 experiment: AdaBoost iteration sweep vs a binary averaged perceptron on the same templates; best held-out result becomes stage-1 and fixes the segmentation gate value.
- **Stage-2 features** (word-level, new; KyTea-global inspired): surface, word length, first/last char and char type, 2-char prefix/suffix, left/right context chars (window 3) and char types, context bigrams. Ablation in Phase 1. Per-word cost: a dozen-odd hash probes × effective class count (1 to a few after masking); at ~1.7 chars/word (ja) this stays well below stage-1 cost per char.
- **Model file**: new headered single-file format (`litsea-two-stage v1` + embedded stage-1 AdaBoost section, lexicon section, stage-2 perceptron section). Both existing loaders were verified to reject the header with `InvalidData`, so existing joint/AdaBoost model files keep working unchanged — the two-stage path is purely additive, selected by model type.
- **API/CLI**: `segment_with_pos()` signature unchanged; new constructor (e.g. `Segmenter::with_two_stage_learner`) added; `--two-stage` flag on extract/train/evaluate; `segment --pos` auto-detects via the model header. Existing `word/POS` corpora feed both stages in one extraction pass.
- **Korean**: stage-1 can use the space-preserving TSV training protocol (#152, 99.91 seg F1), which may substantially *improve* segmentation over joint (80.51); the evaluation-protocol mismatch (spaces removed vs preserved) will be reconciled at the start of Phase 1.

### Gates (finalized from measured variance, not aspirational numbers)

- Throughput (ja, long text): **≥3x** over current ~1.9M chars/s (stretch 5x)
- Segmentation Word F1 (ja): set from the Phase 0 stage-1 experiment ceiling (ballpark 94-95+)
- Tagged F1 (ja): allowed delta vs joint 92.51 set from baseline variance (candidate restriction is expected to help; measure first)
- zh/ko gated per-language with the same methodology

### Phases

- **Phase 0** (premise validation, experiment scripts only): stage-1 quality-recovery experiment; lexicon statistics (single-candidate token rate, held-out unknown rate)
- **Phase 1** (prototype): two-stage inference + training; joint vs two-stage quality/speed on all 3 languages; stage-2 feature ablation
- **Phase 2**: go/no-go with the maintainer
- **Phase 3** (implementation, child issues): model format/loader, runtime, trainer + CLI, retrained bundled models, docs, benches

Risks and mitigations, full feature tables, and per-language details are in the local design doc; this comment is the reviewable summary. Feedback welcome — implementation will not start until this design is approved.

**Comment by mosuka** [comment]
## Phase 0/1 results: premise validation + prototype (quality and speed measured)

All experiments ran against held-out UD GSD test via `litsea evaluate` / replicated span metrics; speed was measured same-process, joint vs two-stage interleaved per round (matched-state A/B). The joint baseline measured 1.4-1.8M chars/s in this harness, consistent with the official 1.83M, so the harness is representative. No repository code was changed — the prototype is a standalone crate using litsea as a path dependency, and stage-2 training reuses the existing `litsea train --pos` on generated word-level feature files.

### Key finding 1: the +5pt segmentation gap was the learner, not the POS signal

A **binary averaged perceptron** (2-class label collapse B-\* → B-X, same 42 templates, existing trainer) essentially reaches joint-level segmentation:

| stage-1 candidate (ja) | seg Word F1 |
|---|---|
| bundled AdaBoost (687 feats) | 91.48 |
| AdaBoost strengthened (t=1e-6, i=50k) | 92.65 — rejected |
| **binary averaged perceptron (58,324 feats)** | **96.48** |
| joint 17-class (reference ceiling) | 96.56 |

Same pattern in all 3 languages: zh 89.62 (joint 90.52, AdaBoost 77.56), ko 81.25 (**beats** joint's 80.51). Collapsing the two class-weight vectors to scalar weights (w = w_B − w_O) is quality-lossless and — notably — **produces a valid file in the existing AdaBoost model format**: it loads and runs on the current engine unchanged, at only ~20% below the bundled model's speed despite 85x more features (confirming #158: cost is family structure, not feature count). Weight pruning: top-30k keeps 96.16, top-20k 95.96.

**Spin-off opportunity**: bundled segmentation models could be upgraded to binary-AP-collapsed models with zero code changes (ja 91.48 → 96.48, zh 77.56 → 89.62), independently of #147.

### Key finding 2: two-stage beats joint on tagging quality

Lexicon statistics validated candidate restriction (gold-tag coverage inside candidate sets: ja 99.09% / zh 98.24% / ko 98.53%; single-candidate tokens ~48-54%; unknown tokens ja 6.1% / zh 12.5% / **ko 34.5%** — the fallback classifier is essential for Korean).

End-to-end (ja; stage-1 = collapsed binary AP, stage-2 = global averaged perceptron with candidate masking + unknown fallback), with feature-set ablation:

| stage-2 feature set (hash probes/scored word) | e2e Tagged F1 | vs joint 92.51 | throughput (bocchan) | ratio vs joint |
|---|---|---|---|---|
| Full (~13) | **93.44** | **+0.93** | 2.1-2.6M c/s | 1.6-2.3x |
| B: −L3/R3/LB/RB (~9) | 93.31 | +0.80 | 3.1-3.2M | 2.2x |
| C: B −L2/R2/type2,3 (~7) | 92.94 | +0.43 | 3.7-4.0M | 2.0-3.2x |
| D: C −TS/FC/LC (~5) | 92.72 | +0.21 | 3.8-4.5M | 2.4-2.8x |

e2e segmentation Word F1 is 96.48 across all sets (fixed by stage-1). **Every set beats the joint model's Tagged F1.** A frequency-dominance cutoff (skip the classifier when one tag covers ≥99% of a surface's training occurrences) cuts scored words 53% → 43% at −0.02pt and is included above.

### Speed verdict vs the 3x gate

Best demonstrated: **2.4-2.8x** (stage-1 collapsed 58k + set D + cutoff 0.99, bocchan corpus, within-round ratios; later rounds showed thermal throttling, so ratios are the reliable metric). The 3x gate is **not yet demonstrated**. Headroom that a production implementation would still add: span-based decoding (no per-word `Vec<String>` allocation), f32/int16 weights, fusing stage-2 into the stage-1 pass, tuned stage-1 pruning + more epochs. Stage-1 alone runs at 5-6x joint, so the architecture's ceiling is well above the gate; the current limiter is ~5 hash probes per scored word plus one surface hash per word.

### Remaining (deferred to implementation phase)

zh/ko stage-2 quality runs (lexicon stats predict ja-like behavior), ko whitespace-protocol reconciliation, per-language gate values, formal benches on an idle machine.

Next: go/no-go decision with the maintainer.

**Comment by mosuka** [comment]
**Go decision (2026-08-14).** Based on the Phase 0/1 results above: quality beats the joint model in every measured configuration (Tagged F1 +0.21 to +0.93, segmentation parity at -0.08pt) with a large model-size reduction, and 2.4-2.8x throughput is demonstrated with known production-side optimizations still unapplied. The speed gate is updated to **>=2.5x required (demonstrated), 3x stretch**. Implementation is split into child issues #166 (model format/loader), #167 (runtime), #168 (training + CLI), #169 (bundled models, cross-language validation, docs, benches). The bundled-segmentation-model spin-off is filed independently as #165.

**Comment by mosuka** [comment]
## Closing summary

All four child issues are merged: #166 (model format, PR #170), #167 (runtime, PR #171), #168 (training pipeline + CLI, PR #172), #169 (bundled models, benches, docs, PR #173). Spin-off #165 (collapsed binary-AP bundled segmentation models) remains open separately.

### Final results (bundled models, `litsea evaluate --pos` + `cargo bench -- external_corpus`, 3 runs)

| Language | Word F1 (joint -> two-stage) | Tagged F1 (joint -> two-stage) | Throughput (joint -> two-stage) |
|---|---|---|---|
| Japanese | 96.56 -> 96.78 | 92.51 -> 92.95 | 1.55M -> 4.38M chars/s (~2.8x, range 2.65-3.05x) |
| Chinese | 90.52 -> 90.82 | 81.18 -> 82.29 | 1.49M -> 3.38M chars/s (~2.3x, range 2.13-2.44x) |
| Korean | 80.51 -> 83.24 | 71.03 -> 78.86 | 2.48M -> 4.54M chars/s (~1.8x, range 1.75-1.90x) |

As bundled, two-stage beats the currently published joint models on both Word F1 and Tagged F1 in every language.

### Acceptance criteria review

- Design doc reviewed before implementation: done (#147 design comment).
- Quality parity gates from measured baseline variance: done — see the epoch-sweep and 3-run bench methodology documented in `docs/src/algorithm/two-stage-tagging.md` (added in #169), including the honest correction of an initial (incorrect) "architectural ceiling" claim for Chinese segmentation.
- Benchmark target (multi-x improvement): met for Japanese (~2.8x, close to the 2.5x-required/3x-stretch target from the go decision); Chinese and Korean land lower (~2.3x, ~1.8x) — per the original design doc's own language ("zh/ko gates set from measured results with the same methodology," not a fixed shared number), this is being reported as the measured outcome rather than treated as a shortfall against an inapplicable target. Korean's lower speedup is explained (34.5% unknown-word rate forces the full stage-2 fallback more often) in the same doc.

### Not removing `--pos` / joint models

Despite two-stage now beating joint on every measured axis as bundled, joint (`--pos`, `with_pos_learner()`, the `*_pos.model` files) is being kept, not deprecated or removed, per this issue's own non-goal ("existing joint-model files must keep working") and because two-stage has no real-world usage yet (all four PRs merged today). This can be revisited as a separate, explicit design decision once two-stage has a track record.
